### PR TITLE
[pack] Selective extension loading for bundles

### PR DESCRIPTION
**Pull request by soninaren**

### CODE REVIEW

File: src/WebJobs.Script.WebHost/FileMonitoringService.cs [hunk 1]
```diff
@@ -162,7 +185,7 @@ private void OnFileChanged(FileSystemEventArgs e)
                 changeDescription = "File";
             }
             else if ((e.ChangeType == WatcherChangeTypes.Deleted || Directory.Exists(e.FullPath))
-                && !_scriptOptions.RootScriptDirectorySnapshot.SequenceEqual(Directory.EnumerateDirectories(_scriptOptions.RootScriptPath)))
+                && _rootDirectorySnapshot.SequenceEqual(Directory.EnumerateDirectories(_scriptOptions.RootScriptPath)))
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/FileMonitoringService.cs]: This logic has been inverted (missing `!`)

File: src/WebJobs.Script.WebHost/Management/WebFunctionsManager.cs [hunk 2]
```diff
@@ -165,7 +166,7 @@ public async Task<(bool, FunctionMetadataResponse)> TryGetFunction(string name,
             // TODO: DI (FACAVAL) Follow up with ahmels - Since loading of function metadata is no longer tied to the script host, we
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Management/WebFunctionsManager.cs]: This change actually addresses the TODO, so that can be removed.

File: src/WebJobs.Script.WebHost/WebScriptHostBuilderExtension.cs [hunk 3]
```diff
@@ -37,7 +38,7 @@ public static class WebScriptHostBuilderExtension
                     services.ConfigureOptions<CustomHttpHeadersOptionsSetup>();
                     services.ConfigureOptions<HostHstsOptionsSetup>();
                 })
-                .AddScriptHost(webHostOptions, configLoggerFactory, webJobsBuilder =>
+                .AddScriptHost(webHostOptions, metadataProvider, configLoggerFactory, webJobsBuilder =>
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/WebScriptHostBuilderExtension.cs]: As discussed, we should be able to refactor this so we're not modifying these extension methods to take additional services.

File: src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs [hunk 4]
```diff
@@ -83,58 +85,63 @@ public async Task<IEnumerable<Type>> GetExtensionsStartupTypesAsync()
...
-                        }
-
-                        string path = item.HintPath;
-                        if (string.IsNullOrEmpty(path))
-                        {
-                            path = assemblyName.Name + ".dll";
-                        }
+            foreach (var extensionItem in extensionItems)
+            {
+                if (extensionItem.Bindings.Count == 0 || extensionItem.Bindings.Intersect(functionBindings, StringComparer.OrdinalIgnoreCase).Count() > 0)
```
Comment (fabiocav) [inline src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs]: Change from `.Count() > 0` to `.Any()`

Comment (fabiocav) [inline src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs]: This logic can only be applied when we're dealing with bundles, as this would be a breaking change for all non-bundle customers.

File: src/WebJobs.Script/Models/ExtensionReference.cs [hunk 5]
```diff
@@ -27,5 +29,10 @@ public class ExtensionReference
         /// implementation.
         /// </summary>
         public string HintPath { get; set; }
+
+        /// <summary>
+        /// Gets the binding exposed by the extension
+        /// </summary>
+        public ICollection<string> Bindings { get; private set; } = new Collection<string>();
```
Comment (fabiocav) [inline src/WebJobs.Script/Models/ExtensionReference.cs]: nit: remove the setter

File: src/WebJobs.Script/ScriptHostBuilderExtensions.cs [hunk 6]
```diff
@@ -38,7 +38,7 @@ namespace Microsoft.Azure.WebJobs.Script
 {
     public static class ScriptHostBuilderExtensions
     {
-        public static IHostBuilder AddScriptHost(this IHostBuilder builder, Action<ScriptApplicationHostOptions> configureOptions, ILoggerFactory loggerFactory = null)
+        public static IHostBuilder AddScriptHost(this IHostBuilder builder, Action<ScriptApplicationHostOptions> configureOptions, IFunctionMetadataProvider functionMetadataProvider, ILoggerFactory loggerFactory = null)
```
Comment (fabiocav) [inline src/WebJobs.Script/ScriptHostBuilderExtensions.cs]: As mentioned above, let's avoid this change.

**Comment by soninaren** [comment]
closes #4223

**Comment by soninaren** [comment]
@fabiocav This is ready for second iteration for review.

File: src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs [hunk 7]
```diff
@@ -83,58 +85,65 @@ public async Task<IEnumerable<Type>> GetExtensionsStartupTypesAsync()
...
-                        }
-
-                        string path = item.HintPath;
-                        if (string.IsNullOrEmpty(path))
-                        {
-                            path = assemblyName.Name + ".dll";
-                        }
+            foreach (var extensionItem in extensionItems)
+            {
+                if (!_extensionBundleManager.IsExtensionBundleConfigured()
```
Comment (fabiocav) [inline src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs]: Cache the results of this lookup

File: src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs [hunk 8]
```diff
@@ -83,58 +85,65 @@ public async Task<IEnumerable<Type>> GetExtensionsStartupTypesAsync()
...
 
-                        return null;
-                    },
-                    (assembly, typeName, ignoreCase) =>
+                    if (extensionType == null)
                     {
-                        _logger.ScriptStartUpLoadedExtension(startupExtensionName, assembly.GetName().Version.ToString());
-                        return assembly?.GetType(typeName, false, ignoreCase);
-                    }, false, true);
+                        _logger.ScriptStartUpUnableToLoadExtension(startupExtensionName, extensionItem.TypeName);
```
Comment (fabiocav) [inline src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs]: nit: remove blank line

File: src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs [hunk 9]
```diff
@@ -83,58 +85,65 @@ public async Task<IEnumerable<Type>> GetExtensionsStartupTypesAsync()
...
+                        continue;
+                    }
+                    if (!typeof(IWebJobsStartup).IsAssignableFrom(extensionType))
+                    {
+                        _logger.ScriptStartUpTypeIsNotValid(extensionItem.TypeName, nameof(IWebJobsStartup));
```
Comment (fabiocav) [inline src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs]: nit: remove blank line

File: src/WebJobs.Script/Diagnostics/Extensions/LoggerExtension.cs [hunk 10]
```diff
@@ -166,6 +166,18 @@ internal static class LoggerExtension
             new EventId(325, nameof(ScriptStartUpLoadedExtension)),
             "Loaded extension '{startupExtensionName}' ({startupExtensionVersion})");
 
+        private static readonly Action<ILogger, Exception> _functionMetadataProviderParsingFunctions =
```
Comment (fabiocav) [inline src/WebJobs.Script/Diagnostics/Extensions/LoggerExtension.cs]: '_functionMetadataProviderParsingFunctions' -> '_functionMetadataProviderReadingMetadata'

File: src/WebJobs.Script/Diagnostics/Extensions/LoggerExtension.cs [hunk 11]
```diff
@@ -166,6 +166,18 @@ internal static class LoggerExtension
...
+            LoggerMessage.Define(
+            LogLevel.Information,
+            new EventId(326, nameof(FunctionMetadataManagerLoadingFunctionsMetadata)),
+            "Reading functions metadata");
+
+        private static readonly Action<ILogger, int, Exception> _functionMetadataProviderFunctionFound =
```
Comment (fabiocav) [inline src/WebJobs.Script/Diagnostics/Extensions/LoggerExtension.cs]: nit: `_functionMetadataProviderFunctionFound` -> `_functionMetadataProviderFunctionsFound`

File: src/WebJobs.Script/Host/FunctionMetadataManager.cs [hunk 12]
```diff
@@ -14,266 +12,46 @@
...
-        public ImmutableDictionary<string, ImmutableArray<string>> Errors
-            => _functionErrors.ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value.ToImmutableArray());
+        public ImmutableDictionary<string, ImmutableArray<string>> Errors { get; private set; }
 
         /// <summary>
         /// Read all functions and populate function metadata.
         /// </summary>
         private ImmutableArray<FunctionMetadata> LoadFunctionMetadata()
         {
+            var functionsWhiteList = _scriptOptions.Value.Functions;
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/FunctionMetadataManager.cs]: Here and on 44; change from `var` to type

File: src/WebJobs.Script/Host/FunctionMetadataProvider.cs [hunk 13]
```diff
@@ -0,0 +1,264 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/FunctionMetadataProvider.cs]: Can you confirm there's no new logic here? This is only what was moved from the metadata manager, correct?

File: test/WebJobs.Script.Tests.Integration/TestFunctionHost.cs [hunk 14]
```diff
@@ -283,12 +288,34 @@ public void ConfigureServices(IServiceCollection services)
...
             }
 
-            public void Configure(AspNetCore.Builder.IApplicationBuilder app, AspNetCore.Hosting.IApplicationLifetime applicationLifetime, AspNetCore.Hosting.IHostingEnvironment env, ILoggerFactory loggerFactory)
+            public void Configure(AspNetCore.Builder.IApplicationBuilder app, AspNetCore.Hosting.IApplicationLifetime applicationLifetime, AspNetCore.Hosting.IHostingEnvironment env, Microsoft.Extensions.Logging.ILoggerFactory loggerFactory)
             {
                 _startup.Configure(app, applicationLifetime, env, loggerFactory);
             }
         }
 
+        private class TestServiceProvider : IServiceProvider
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests.Integration/TestFunctionHost.cs]: This shouldn't be needed... you should be able to build a new service provider with the appropriate services. There are examples of that in the test suite.

File: test/WebJobs.Script.Tests.Shared/TestHelpers.cs [hunk 15]
```diff
@@ -35,7 +35,9 @@ public static string FunctionsTestDirectory
         {
             get
             {
-                return Path.Combine(Path.GetTempPath(), "FunctionsTest");
+                var path = Path.Combine(Path.GetTempPath(), "FunctionsTest");
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests.Shared/TestHelpers.cs]: This does change the behavior of this helper method. Are you sure there's no existing code that expects to test the result of this property? You may also want to consider changing this to a method and make its behavior clear in the name (as it isn't after this change). That could have a larger impact on the changes. Adding a method to this type that introduces this logic would be a better option.

File: test/WebJobs.Script.Tests.Shared/TestHostBuilderExtensions.cs [hunk 16]
```diff
@@ -78,12 +88,41 @@ public static IHostBuilder ConfigureDefaultTestWebScriptHost(this IHostBuilder b
...
+            var workerOptions = new LanguageWorkerOptions
+            {
+                WorkerConfigs = TestHelpers.GetTestWorkerConfigs()
+            };
+            var metadataProvider = new FunctionMetadataProvider(optionsMonitor, new OptionsWrapper<LanguageWorkerOptions>(workerOptions), NullLogger<FunctionMetadataProvider>.Instance);
+            _metadataProvider = metadataProvider;
+            return services.AddSingleton<IFunctionMetadataProvider>(metadataProvider);
         }
+
+        //private class TestServiceProvider : IServiceProvider
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests.Shared/TestHostBuilderExtensions.cs]: nit: Commented out code....

File: test/WebJobs.Script.Tests.Shared/TestHostBuilderExtensions.cs [hunk 17]
```diff
@@ -2,25 +2,34 @@
...
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Options;
 using Moq;
+using FunctionMetadata = Microsoft.Azure.WebJobs.Script.Description.FunctionMetadata;
 
 namespace Microsoft.WebJobs.Script.Tests
 {
     public static class TestHostBuilderExtensions
     {
+        private static IFunctionMetadataProvider _metadataProvider;
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests.Shared/TestHostBuilderExtensions.cs]: This doesn't seem to be used.

File: test/WebJobs.Script.Tests/Configuration/DefaultDependencyValidatorTests.cs [hunk 18]
```diff
@@ -87,7 +87,7 @@ private async Task<LogMessage> RunTest(Action<IServiceCollection> configureWebHo
                         }));
 
                         string uniqueTestRootPath = Path.Combine(Path.GetTempPath(), "FunctionsTest", "DependencyValidatorTests");
-
+                        Directory.CreateDirectory(uniqueTestRootPath);
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests/Configuration/DefaultDependencyValidatorTests.cs]: Why is this change required?

File: test/WebJobs.Script.Tests/FileMonitoringServiceTests.cs [hunk 19]
```diff
@@ -27,6 +27,63 @@ public static void GetRelativeDirectory_ReturnsExpectedDirectoryName(string path
...
+                };
+                var loggerFactory = new LoggerFactory();
+                var mockWebHostEnvironment = new Mock<IScriptJobHostEnvironment>(MockBehavior.Loose);
+                var mockEventManager = new ScriptEventManager();
+
+                // Act
+                FileMonitoringService fileMonitoringService = new FileMonitoringService(new OptionsWrapper<ScriptJobHostOptions>(jobHostOptions),
+                    loggerFactory, mockEventManager, mockWebHostEnvironment.Object);
+                Assert.Equal(fileMonitoringService.GetDirectorySnapshot().Length, 1);
+                Assert.Equal(fileMonitoringService.GetDirectorySnapshot()[0], tempDir + "\\Host");
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests/FileMonitoringServiceTests.cs]: This concatenation will have issues on Linux

File: test/WebJobs.Script.Tests/FunctionMetadataProviderTests.cs [hunk 20]
```diff
@@ -214,24 +196,14 @@ public void ValidateFunctionName_DoesNotThrowOnValidName(string functionName)
...
+                FunctionMetadataProvider.ValidateName(functionName);
             }
             catch (InvalidOperationException)
             {
                 Assert.True(false, $"Valid function name {functionName} failed validation.");
             }
         }
 
-        [Fact]
-        public void ReadFunctionMetadata_Succeeds()
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests/FunctionMetadataProviderTests.cs]: Where is this test now?

File: test/WebJobs.Script.Tests/Managment/FunctionsSyncManagerTests.cs [hunk 21]
```diff
@@ -11,12 +12,15 @@
 using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
+using Castle.Core.Logging;
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests/Managment/FunctionsSyncManagerTests.cs]: Is this correct?

File: test/WebJobs.Script.Tests/Managment/WebFunctionsManagerTests.cs [hunk 22]
```diff
@@ -3,13 +3,15 @@
...
 using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.IO;
 using System.IO.Abstractions;
 using System.Linq;
 using System.Net.Http;
 using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
+using Castle.Core.Logging;
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests/Managment/WebFunctionsManagerTests.cs]: Is this correct?

File: test/WebJobs.Script.Tests/ScriptHostTests.cs [hunk 23]
```diff
@@ -337,6 +337,7 @@ public async Task Initialize_WithInvalidSiteExtensionVersion_Throws(string exten
             try
             {
                 string rootPath = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
+                Directory.CreateDirectory(rootPath);
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests/ScriptHostTests.cs]: Looks like some behavior changed that forced you to create this throughout the tests, can you clarify? Were we able to cope with this before but have an issue now?

File: test/WebJobs.Script.Tests/ScriptStartupTypeDiscovererTests.cs [hunk 24]
```diff
@@ -165,7 +177,10 @@ public async Task GetExtensionsStartupTypes_UnableToDownloadExtensionBundle_Retu
             LoggerFactory factory = new LoggerFactory();
             factory.AddProvider(testLoggerProvider);
             var testLogger = factory.CreateLogger<ScriptStartupTypeLocator>();
-            var discoverer = new ScriptStartupTypeLocator(string.Empty, testLogger, mockExtensionBundleManager.Object);
+
+            var functionMetadataProvider = new Mock<IFunctionMetadataProvider>();
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests/ScriptStartupTypeDiscovererTests.cs]: Can we add a method to avoid this duplication through this class?

File: src/WebJobs.Script/Host/FunctionMetadataProvider.cs (hunk 13 above)
Comment (soninaren) [reply src/WebJobs.Script/Host/FunctionMetadataProvider.cs]: Yes everything here was moved from metadata manager. Everything in here is the same. Except for the following 2 things, which we already covered in the previous iteration.
1. Function filtering based on FunctionWhiteList in host.json is still in function metdatamanager
2. Static method from metadataManager has been converted to instance method. WebFunctionManager instead of using static method now uses cached function from metadata from provider.

File: test/WebJobs.Script.Tests/Managment/FunctionsSyncManagerTests.cs (hunk 21 above)
Comment (soninaren) [reply test/WebJobs.Script.Tests/Managment/FunctionsSyncManagerTests.cs]: removed.

File: test/WebJobs.Script.Tests/Managment/WebFunctionsManagerTests.cs (hunk 22 above)
Comment (soninaren) [reply test/WebJobs.Script.Tests/Managment/WebFunctionsManagerTests.cs]: removed

File: test/WebJobs.Script.Tests/ScriptHostTests.cs (hunk 23 above)
Comment (soninaren) [reply test/WebJobs.Script.Tests/ScriptHostTests.cs]: The reason i had to add it is because the tests did not expect the host to read the function metadata before. Because of this it did not matter if the rootpath existed or not. But now that we have moved up the logic to read the metadata the tests invariably do need the path to be present. I was avoiding any change to the read metadata logic of the function. But i had updated it to return an empty collection of function metadata in case the rootpath does not exist.

File: test/WebJobs.Script.Tests.Shared/TestHelpers.cs (hunk 15 above)
Comment (soninaren) [reply test/WebJobs.Script.Tests.Shared/TestHelpers.cs]: Have reverted this and updated the logic in the metadata provider to be a bit more resilient

File: test/WebJobs.Script.Tests/Configuration/DefaultDependencyValidatorTests.cs (hunk 18 above)
Comment (soninaren) [reply test/WebJobs.Script.Tests/Configuration/DefaultDependencyValidatorTests.cs]: As mentioned in the comment below. The tests did not expect the host to read the function metadata before. Because of this it did not matter if the rootpath existed or not. But now that we have moved up the logic to read the metadata the tests invariably do need the path to be present.

File: test/WebJobs.Script.Tests/FunctionMetadataProviderTests.cs (hunk 20 above)
Comment (soninaren) [reply test/WebJobs.Script.Tests/FunctionMetadataProviderTests.cs]: I had missed this one. Have added it back here.
https://github.com/Azure/azure-functions-host/pull/4980/files#diff-210d2369bbeb2645e900222fa323617aR43

**Comment by soninaren** [comment]
@fabiocav Ping. Just need final validation on this.

File: src/WebJobs.Script.WebHost/FileMonitoringService.cs [hunk 25]
```diff
@@ -54,6 +56,23 @@ public FileMonitoringService(IOptions<ScriptJobHostOptions> scriptOptions, ILogg
...
+        {
+            if (_scriptOptions.RootScriptPath != null)
+            {
+                try
+                {
+                    return Directory.EnumerateDirectories(_scriptOptions.RootScriptPath).ToImmutableArray();
+                }
+                catch (DirectoryNotFoundException)
+                {
+                    return ImmutableArray<string>.Empty;
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/FileMonitoringService.cs]: This can be removed... just catch (log perhaps) and let the execution flow. No need for the duplication here

File: src/WebJobs.Script/Host/FunctionMetadataManager.cs [hunk 26]
```diff
@@ -14,266 +12,46 @@
...
-                case "dll":
-                    return DotNetScriptTypes.DotNetAssembly;
-            }
-            var workerConfig = workerConfigs.FirstOrDefault(config => config.Extensions.Contains("." + extension));
-            if (workerConfig != null)
-            {
-                return workerConfig.Language;
+                _logger.LogInformation($"A function whitelist has been specified, excluding all but the following functions: [{string.Join(", ", functionsWhiteList)}]");
+                metadata = metadata.Where(function => functionsWhiteList.Any(functionName => functionName.Equals(function.Name, StringComparison.CurrentCultureIgnoreCase))).ToImmutableArray();
+                Errors = _functionMetadataProvider.FunctionErrors.Where(kvp => functionsWhiteList.Contains(kvp.Key)).ToImmutableDictionary<string, ImmutableArray<string>>();
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/FunctionMetadataManager.cs]: The code above performs a case insensitive comparison using current culture, this doesn’t. Should these be consistent?

File: test/WebJobs.Script.Tests/FileMonitoringServiceTests.cs [hunk 27]
```diff
@@ -27,6 +27,63 @@ public static void GetRelativeDirectory_ReturnsExpectedDirectoryName(string path
...
+                Assert.False(fileMonitoringService.GetDirectorySnapshot().IsDefault);
+                Assert.True(fileMonitoringService.GetDirectorySnapshot().IsEmpty);
+            }
+        }
+
+        [Fact]
+        public void InitializesGetDirectorySnapshot()
+        {
+            using (var directory = new TempDirectory())
+            {
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests/FileMonitoringServiceTests.cs]: Most of this method is identical to what you have above. Can we consolidate this logic to avoid duplication?

File: test/WebJobs.Script.Tests/ScriptHostTests.cs [hunk 28]
```diff
@@ -369,6 +369,7 @@ public async Task Initialize_WithLatestSiteExtensionVersion_LogsWarning()
...
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests/ScriptHostTests.cs]: Use the test/temp directory helpers here to ensure this is cleaned up

File: src/WebJobs.Script/Host/FunctionMetadataManager.cs (hunk 26 above)
Comment (soninaren) [reply src/WebJobs.Script/Host/FunctionMetadataManager.cs]: You are right we should do case-insensitive comparison here too

File: src/WebJobs.Script/Host/FunctionMetadataManager.cs [hunk 29]
```diff
@@ -14,266 +12,46 @@
...
-                case "dll":
-                    return DotNetScriptTypes.DotNetAssembly;
-            }
-            var workerConfig = workerConfigs.FirstOrDefault(config => config.Extensions.Contains("." + extension));
-            if (workerConfig != null)
-            {
-                return workerConfig.Language;
+                _logger.LogInformation($"A function whitelist has been specified, excluding all but the following functions: [{string.Join(", ", functionsWhiteList)}]");
+                metadata = metadata.Where(function => functionsWhiteList.Any(functionName => functionName.Equals(function.Name, StringComparison.CurrentCultureIgnoreCase))).ToImmutableArray();
+                Errors = _functionMetadataProvider.FunctionErrors.Where(kvp => functionsWhiteList.Any(functionName => functionName.Equals(kvp.Key))).ToImmutableDictionary<string, ImmutableArray<string>>();
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/FunctionMetadataManager.cs]: I think we missed the comparison change...
